pennylane_lightning/lightning_gpu/_state_vector.py

```diff
@@ -35,13 +68,303 @@ class LightningGPUStateVector(LightningBaseStateVector):
+        Args:
+            index (int): integer representing the computational basis state.
+            use_async(bool): indicates whether to use asynchronous memory copy from host to device or not.
+            Note: This function only supports synchronized memory copy.
+        """
+        self._qubit_state.setBasisState(index, use_async)
+
+    def _apply_state_vector(self, state, device_wires, use_async=False):
+        """Initialize the state vector on GPU with a specified state on host.
+        Note that any use of this method will introduce host-overheads.
+        Args:
+        state (array[complex]): normalized input state (on host) of length ``2**len(wires)``
+                or broadcasted state of shape ``(batch_size, 2**len(wires))``
+        device_wires (Wires): wires that get initialized in the state
+        use_async(bool): indicates whether to use asynchronous memory copy from host to device or not.
+        Note: This function only supports synchronized memory copy from host to device.
+        """
+
+        if isinstance(state, self._qubit_state.__class__):
+            raise DeviceError("LightningGPU does not support allocate external state_vector.")
+
+            # TODO
+            # state_data = allocate_aligned_array(state.size, np.dtype(self.dtype), True)
+            # state.getState(state_data)
+            # state = state_data
+
+        state = self._asarray(state, dtype=self.dtype)  # this operation on host
+        output_shape = [2] * self._num_local_wires
+
+        if len(device_wires) == self.num_wires and Wires(sorted(device_wires)) == device_wires:
+            # Initialize the entire device state with the input state
+            if self.num_wires == self._num_local_wires:
+                self.syncH2D(np.reshape(state, output_shape))
+                return
+            local_state = np.zeros(1 << self._num_local_wires, dtype=self.C_DTYPE)
+            self._mpi_handler.mpi_manager.Scatter(state, local_state, 0)
+            self.syncH2D(np.reshape(local_state, output_shape))
+            return
+
+        # generate basis states on subset of qubits via the cartesian product
+        basis_states = np.array(list(product([0, 1], repeat=len(device_wires))))
+
+        # get basis states to alter on full set of qubits
+        unravelled_indices = np.zeros((1 << len(device_wires), self.num_wires), dtype=int)
+        unravelled_indices[:, device_wires] = basis_states
+
+        # get indices for which the state is changed to input state vector elements
+        ravelled_indices = np.ravel_multi_index(unravelled_indices.T, [2] * self.num_wires)
+
+        # set the state vector on GPU with the unravelled_indices and their corresponding values
+        self._qubit_state.setStateVector(
+            ravelled_indices, state, use_async
+        )  # this operation on device
+
+    def _apply_basis_state(self, state, wires):
```

Comment:
Then again, implementing a non abstractmethod! We should be able to unify the implementations across Lightning devices.

Reply:
Unified!